### ISSUE TITLE
Wrapping a serverless function in `useRequireAuth` still allows you to execute the function without being authenticated

### ISSUE THREAD

**Issue by Tobbe** [redwoodjs/graphql#5198]
We had a user on Discord who was wrapping his serverless function in `useRequireAuth`, but could still execute the function without any auth headers set. He was expecting that `useRequireAuth` would block any unauthenticated invocations. Which is totally understandable, as the documentation isn't really clear here.

What `useRequireAuth` really does though, is it make all the other auth related functions available. Maybe it should really be called `makeAuthAvailable()`? Or `withAuthAvailable()` to copy React HOC naming conventions.

What you really need to do to block execution is

```
const myHandler = async (_event: APIGatewayEvent, _context: Context) => {
  requireAuth({ roles: 'admin' })

  return {
    statusCode: 200,
  }
}

export const handler = useRequireAuth({ handlerFn: myHandler, getCurrentUser })
```

So, wrap the handler in `useRequireAuth()` and then call `requireAuth()` first thing in your handler.

I do think the current implementation is the right one. We could block execution in `useRequireAuth`, but doing the current way is much more flexible. For example it allows the developer to have two execution paths through the handler - one for authenticated users and one for unauthenticated users.

@dthyresson Do you think we should just update the docs, or should we rename something? Or both? Or something else entirely?

**Comment by dthyresson** [comment]
@Tobbe Many thanks for looking into this issue.

Now that I think back on it, 

```
const myHandler = async (_event: APIGatewayEvent, _context: Context) => {
  requireAuth({ roles: 'admin' })

  return {
    statusCode: 200,
  }
}

export const handler = useRequireAuth({ handlerFn: myHandler, getCurrentUser })
```

Is the implementation that @dac09 and I intended, but I must have documented it incorrectly and forgot.

As you said, the wrapper "makes auth available" and then you keep to "use" it.

Hence, here you can enforce role access.

I think we simple need to update the docs -- @dac09, agree?

**Comment by Tobbe** [comment]
With a function name like `useRequireAuth` it's not difficult to understand how someone might think it'll internally use requireAuth :)  

Looking at the terminology you've used for some of the tests you've written, maybe a better name would have been `withAuthEnrichment()`. We could have both names for a while and deprecate `useRequireAuth`. That way it wouldn't be a breaking change (yet).

**Comment by vikotar** [comment]
Thanks for the help guys. I was able to get my authentication working. 😎

**Comment by jtoar** [comment]
We discussed this one a bit today; the conclusion was just that it needs more async discussion, and @dthyresson is now tagged on it. I think we were leaning more towards a name change of some sort now, but not confirmed.

**Comment by Tobbe** [comment]
Documentation updates are definitely needed. But as @jtoar said, I also got a feeling we were leaning towards **also** considering a name change. Still looking for @dac09's input as well.

**Comment by dac09** [comment]
Yeah agreed - we can change the name, and reexport to make it non-breaking. We can discuss on the CTM!

**Comment by Tobbe** [comment]
Is `withAuthEnrichment()` a good name? Any other suggestions?

**Comment by dac09** [comment]
I’d wait for Tom’s suggestions!

**Comment by jtoar** [comment]
@Tobbe I like `withAuth`, but not the `Enrichment` part so much. 

The confusion around `useRequireAuth` is mainly that it sounds like a React hook. If I can stretch the React metaphor a bit, what `useRequireAuth` does right now is more like a React context provider: it makes auth "available". But you still have to use it (via `requireAuth`).

What are providers normally called in react? `SomethingProvider` right? That still sounds weird as a function. "with" sounds a little better. But TBD!

**Comment by Tobbe** [comment]
@jtoar Thanks for your input. "Enriched" is used a bit in the tests for `useRequireAuth`. That's where I got that from. But definitely open to other names. 

https://github.com/redwoodjs/redwood/blob/main/packages/graphql-server/src/functions/__tests__/useRequireAuth.test.ts#L141

**Comment by Tobbe** [comment]
We decided on a call today to call the function `withAuth`
